client/consul.go

```diff
@@ -0,0 +1,170 @@
+	consulConf := conf.Consul
+	transport := hcat.TransportInput{
+		SSLEnabled: *consulConf.TLS.Enabled,
+		SSLVerify:  *consulConf.TLS.Verify,
+		SSLCert:    *consulConf.TLS.Cert,
+		SSLKey:     *consulConf.TLS.Key,
+		SSLCACert:  *consulConf.TLS.CACert,
+		SSLCAPath:  *consulConf.TLS.CAPath,
+		ServerName: *consulConf.TLS.ServerName,
+
+		DialKeepAlive:       *consulConf.Transport.DialKeepAlive,
+		DialTimeout:         *consulConf.Transport.DialTimeout,
+		DisableKeepAlives:   *consulConf.Transport.DisableKeepAlives,
+		IdleConnTimeout:     *consulConf.Transport.IdleConnTimeout,
+		MaxIdleConns:        *consulConf.Transport.MaxIdleConns,
+		MaxIdleConnsPerHost: *consulConf.Transport.MaxIdleConnsPerHost,
+		TLSHandshakeTimeout: *consulConf.Transport.TLSHandshakeTimeout,
+	}
+
+	consul := hcat.ConsulInput{
+		Address:      *consulConf.Address,
+		Token:        *consulConf.Token,
+		AuthEnabled:  *consulConf.Auth.Enabled,
+		AuthUsername: *consulConf.Auth.Username,
+		AuthPassword: *consulConf.Auth.Password,
+		Transport:    transport,
+	}
+
+	clients := hcat.NewClientSet()
+	if err := clients.AddConsul(consul); err != nil {
+		return nil, err
+	}
+
+	logger := logging.Global().Named(loggingSystemName).Named(consulSubsystemName)
+
+	r := retry.NewRetry(maxRetry, time.Now().UnixNano())
+	c := &ConsulClient{
+		Client: clients.Consul(),
+		retry:  r,
+		logger: logger}
+	return c, nil
+}
+
+// GetLicense queries Consul for a signed license, and returns it if available
+func (c *ConsulClient) GetLicense(ctx context.Context, q *consulapi.QueryOptions) (string, error) {
+	c.logger.Debug("getting license")
+
+	desc := "consul client get license"
+	var err error
+	var license string
+
+	f := func(context.Context) error {
+		license, err = c.Operator().LicenseGetSigned(q)
+		if err != nil {
+			license = ""
+			return err
+		}
+		return nil
+	}
+
+	err = c.retry.Do(ctx, f, desc)
+
+	return license, err
+}
+
+// IsEnterprise queries Consul for information about itself, it then
+// parses this information to determine whether the Consul being
+// queried is Enterprise or OSS. Returns true if Consul is Enterprise.
+func (c *ConsulClient) IsEnterprise(ctx context.Context) (bool, error) {
+	c.logger.Debug("getting sku")
+
+	desc := "consul client get sku"
+	var err error
+	var info ConsulAgentConfig
+
+	f := func(context.Context) error {
+		info, err = c.Agent().Self()
+		if err != nil {
+			info = nil
+			return err
+		}
+		return nil
+	}
+
+	err = c.retry.Do(ctx, f, desc)
+	if err != nil {
+		return false, err
+	}
+
+	ctx = logging.WithContext(ctx, c.logger)
+
+	isEnterprise, err := isConsulEnterprise(ctx, info)
+	if err != nil {
+		return false, fmt.Errorf("unable to parse sku: %v", err)
+	}
+	return isEnterprise, nil
+}
+
+func isConsulEnterprise(ctx context.Context, info ConsulAgentConfig) (bool, error) {
+	logger := logging.FromContext(ctx)
+	v, ok := info["Config"]["Version"].(string)
```

Comment:
@jwair this is an interesting goism that John brought to my attention last week. This will actually return false if the keys don't exist and if the casting to `string` were to fail